File: javascript/leetcode/combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(c, target) {
    if (target < 0) {
        return [];
    }
    if (target == 0) {
        return [[]];
    }
    var result = []
    for (var i = 0; i < c.length; i++) {
        var partSum = combinationSum(c.slice(i), target - c[i])
        partSum.forEach(function(s) {
            result.push(s.concat(c[i])) 
        });
    }
    return result
};

File: ruby/leetcode/combination-sum.rb
# @param {Integer[]} candidates
# @param {Integer} target
# @return {Integer[][]}
def combination_sum(candidates, target)
    return [] if target < 0
    return [[]] if target == 0
    result = []
    for i in 0...candidates.size()
        part_sum = combination_sum(candidates[i..-1], target - candidates[i])
        part_sum.each do |s|
            result << s + [candidates[i]]
        end
    end
    return result

end
